sycl/include/CL/sycl/detail/usm_dispatch.hpp

```diff
@@ -0,0 +1,73 @@
+//==-------------- usm_dispatch.hpp - SYCL USM Dispatch --------*- C++ -*---==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// ===--------------------------------------------------------------------=== //
+#pragma once
+
+#include <CL/sycl/detail/clusm.hpp>
+
+#include <memory>
+
+namespace cl {
+namespace sycl {
+namespace detail {
+namespace usm {
+
+class USMDispatcher {
+public:
+  USMDispatcher(cl_platform_id Platform);
+  
+  void *hostMemAlloc(pi_context Context, cl_mem_properties_intel *Properties,
+                     size_t Size, pi_uint32 Alignment, pi_result *ErrcodeRet);
+  void *deviceMemAlloc(pi_context Context, pi_device Device,
+                       cl_mem_properties_intel *Properties, size_t Size,
+                       pi_uint32 Alignment, pi_result *ErrcodeRet);
+  void *sharedMemAlloc(pi_context Context, pi_device Device,
+                       cl_mem_properties_intel *Properties, size_t Size,
+                       pi_uint32 Alignment, pi_result *ErrcodeRet);
+  pi_result memFree(pi_context Context, void *Ptr);
+  pi_result setKernelArgMemPointer(pi_kernel Kernel, pi_uint32 ArgIndex,
+                                   const void *ArgValue);
+  void setKernelIndirectAccess(pi_kernel Kernel, pi_queue Queue);
+  pi_result enqueueMemset(pi_queue Queue, void *Ptr, pi_int32 Value,
+                          size_t Count, pi_uint32 NumEventsInWaitList,
+                          const pi_event *EventWaitList, pi_event *Event);
+  pi_result enqueueMemcpy(pi_queue Queue, pi_bool Blocking, void *DestPtr,
+                          const void *SrcPtr, size_t Size,
+                          pi_uint32 NumEventsInWaitList,
+                          const pi_event *EventWaitList, pi_event *Event);
+  pi_result enqueueMigrateMem(pi_queue Queue, const void *Ptr, size_t Size,
+                              cl_mem_migration_flags Flags,
+                              pi_uint32 NumEventsInWaitList,
+                              const pi_event *EventWaitList, pi_event *Event);
+  pi_result enqueueMemAdvise(pi_queue Queue, void *Ptr, size_t Size,
+                             cl_mem_advice_intel Advice,
+                             pi_uint32 NumEventsInWaitList,
+                             const pi_event *EventWaitList, pi_event *Event);
+  pi_result getMemAllocInfo(pi_context Context, const void *Ptr,
+                            cl_mem_info_intel ParamName, size_t ParamValueSize,
+                            void *ParamValue, size_t *ParamValueSizeRet);
+
+private:
+  bool mEmulated = false;
```

Comment:
Probably instead of having this var you can check if mEmulator is nullptr or not.